### PR TITLE
Implement limit and order by clauses and update to `1.0.0-pre1`

### PR DESCRIPTION
**Pull request by mdrachuk**
This PR adds the notion of `Query` `Pagination` and `OrderBy` clause to Datastore storage implementation.

The library version and the dependencies to `core-java` and `base` are set to `1.0.0-pre1`.

### CODE REVIEW

File: datastore/src/main/java/io/spine/server/storage/datastore/DsEntityComparator.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        private final ValueType valueType;
+
+        ComparableValueExtractor(ValueType type) {
+            this.valueType = type;
+        }
+
+        abstract @Nullable Comparable extract(Value<?> value);
+
+        private boolean matches(ValueType type) {
+            return type == valueType;
```
[the remark this answers is not in the document]
Comment (mdrachuk) [reply datastore/src/main/java/io/spine/server/storage/datastore/DsEntityComparator.java]: Does not apply: ValueType is an enum.

**Comment by mdrachuk** [comment]
@armiol, PTAL. I still need to fix code coverage, but this won’t change the existing code.

File: .travis.yml [hunk 2]
```diff
@@ -5,8 +5,8 @@ jdk:
...
-
+  - secure: N30+tQHzr7i05iIbVf6YAk7i3ykPjZLxEZur3cZWh5vwD4TmQI9FDZ5vVJo3HWulb8tCV5jFM0RXwo2nswhq2G8WGJY6DkjICF1/U3FPax2VdgDEMScfULvRbt5f6NugHo5sbUuVPukVKv4+7LBdOuBOFXQ92OfEm9HQVO6PSzg5jgHeSJSEmh1cW+Zn6fSRVBP7+1A1HO/i0zdccA4noxx+pbhWp28GM7z2jnQ5x21Up9Zs4HA599r3pTYQuguzeaUzJh9GXMRh6xio1FgDfQ2wJe/0cTR+ZXcZNefX9LKo+eKNeez3ylgxkq48THtIfko5I7DBnm8omGbMSgoX1HF8CVgBD8igHUjN7nk+xbIR6aJLGaoZfdwxb/th2jayoj6JNp2v21KbmoTWlmUnOx5tFXAJIf1Y8+tl6PEfPmlz+LCXug6xhLTg67yD4Ee2EjMje9AUsbOjc033PMgszMPajjN5+m497ys7d7v+qU/6WeXh4OKGg04jkFw0hAghYd/+0twe9iUgPaFTm+pfmOKkOQjj9g8mhwwqQ57OdPzqDBl7EWNjHQtMgF0whWfitWzYUaqHUdRke4M74G9Albr8iK3IUXGBf/UKj8xWYF92J6TdgkNuMT1CR4XaYdYqyu4B6fjvCHXfaukt8G00y4cnY7OPTd4JEdwwTPSBvHQ=
```
Comment (armiol) [inline .travis.yml]: Please comment on purpose.

File: build.gradle [hunk 3]
```diff
@@ -47,13 +49,16 @@ apply from: 'version.gradle'
...
     runsOnWindows = org.gradle.internal.os.OperatingSystem.current().isWindows()
+    spineProtobufPluginId = 'io.spine.tools.spine-model-compiler'
 
     projectsToPublish = ['datastore']
 }
 
 allprojects {
     apply plugin: 'maven'
     apply plugin: 'jacoco'
+    apply plugin: 'idea'
```
Comment (armiol) [inline build.gradle]: I don't think we need that for the root project. So consider moving this to `subprojects`.

File: datastore/src/main/java/io/spine/server/storage/datastore/DatastoreWrapper.java [hunk 4]
```diff
@@ -197,50 +198,149 @@ public Entity read(Key key) {
      * <p>The resulting {@code Iterator} is evaluated lazily. A call to
      * {@link Iterator#remove() Iterator.remove()} causes an {@link UnsupportedOperationException}.
      *
-     * @param keys {@link Key Keys} to search for
+     * <p>The results are returned in an order matching that of the provided keys with nulls
```
Comment (armiol) [inline datastore/src/main/java/io/spine/server/storage/datastore/DatastoreWrapper.java]: `{@code null}s`

File: datastore/src/main/java/io/spine/server/storage/datastore/DatastoreWrapper.java [hunk 5]
```diff
@@ -197,50 +198,149 @@ public Entity read(Key key) {
...
-            result = readBulk(keysList);
+        return keysList.size() <= MAX_KEYS_PER_READ_REQUEST
+               ? actor.get(toArray(keys, Key.class))
+               : readBulk(keysList);
+    }
+
+    private static Iterator<@Nullable Entity> orderByKeys(Iterable<Key> keys,
+                                                          Iterator<Entity> items) {
+        List<Entity> entities = newLinkedList(() -> items);
+        return stream(keys)
```
Comment (armiol) [inline datastore/src/main/java/io/spine/server/storage/datastore/DatastoreWrapper.java]: Please extract the returning value into a variable.

File: datastore/src/main/java/io/spine/server/storage/datastore/DatastoreWrapper.java [hunk 6]
```diff
@@ -197,50 +198,149 @@ public Entity read(Key key) {
...
+     * <p>The resulting {@code Iterator} is evaluated lazily. A call to
+     * {@link Iterator#remove() Iterator.remove()} causes an {@link UnsupportedOperationException}.
+     *
+     * @param query
+     *         {@link Query} to execute upon the Datastore
+     * @return results fo the query as a lazily evaluated {@link Iterator}
+     * @throws IllegalArgumentException
+     *         if the provided {@linkplain StructuredQuery#getLimit() query includes a limit}
+     */
+    private Iterator<Entity> readAllInBatches(StructuredQuery<Entity> query,
```
Comment (armiol) [inline datastore/src/main/java/io/spine/server/storage/datastore/DatastoreWrapper.java]: Let's name it `readAll`. The `batchSize` parameter tells us about the batches.

File: datastore/src/main/java/io/spine/server/storage/datastore/DsEntityComparator.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        Comparable bValue = ComparableValueExtractor.comparable(b.getValue(column));
+
+        if (aValue == null) {
+            return -1;
+        }
+        if (bValue == null) {
+            return +1;
+        }
+
+        //noinspection unchecked entity values are required to be comparable by Spine
```
Comment (armiol) [inline datastore/src/main/java/io/spine/server/storage/datastore/DsEntityComparator.java]: Please fix this `noinspection`.

File: datastore/src/main/java/io/spine/server/storage/datastore/DsLookupByIds.java [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+     * @param orderBy
+     *         a specification of order in which the query results must be returned
+     * @return an iterator over the matching entity records
+     */
+    Iterator<EntityRecord> find(Iterable<I> ids,
+                                FieldMask fieldMask,
+                                Predicate<Entity> predicate,
+                                OrderBy orderBy,
+                                long limit) {
+        return read(ids)
```
Comment (armiol) [inline datastore/src/main/java/io/spine/server/storage/datastore/DsLookupByIds.java]: Please extract a variable before returning the value.

File: datastore/src/main/java/io/spine/server/storage/datastore/DsLookupByIds.java [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+     *         a predicate which must be matched by entities to be returned as results
+     * @param orderBy
+     *         a specification of order in which the query results must be returned
+     * @return an iterator over the matching entity records
+     */
+    Iterator<EntityRecord> find(Iterable<I> ids,
+                                FieldMask fieldMask,
+                                Predicate<Entity> predicate,
+                                OrderBy orderBy) {
+        return read(ids)
```
Comment (armiol) [inline datastore/src/main/java/io/spine/server/storage/datastore/DsLookupByIds.java]: Same as above.

I also don't understand why this method and the method above do not share any code. Please review this method and all the similar methods below for returning the variable (useful for inspecting in debug) and for the code duplication.

**Review by armiol — CHANGES_REQUESTED** [review]
@mdrachuk please see my comments.

**Review by armiol — APPROVED** [review]

**Pull request merged**